Inference Cost proposal (#1646)
llm-d/llm-d: docs/proposals/cost-metrics-primer.md

### Usage-only cost

Each workload pays only for what it actually consumed. App-a pays for 1 CPU, same as app-b. The wasted 2 CPUs (app-a's over-provisioning) now join the idle pool as unattributed cost. **The numbers no longer add up to your overall bill.**

### The practical difference - usage vs allocation based costs

|                              | Allocation-based                              | Usage-only                                          |
|------------------------------|-----------------------------------------------|-----------------------------------------------------|
| Reconciles to cloud bill     | Yes                                           | No — waste + idle are unattributed                  |                                             |
| Incentivizes right-sizing    | Yes — you pay for what you reserve            | Less so                                             |

> | Identifies over-provisioning | Not directly — baked into workload cost       | Yes — visible as the gap between the two cost figures |

Reviewer: typo:
```suggestion
Identifies over-provisioning | Not directly — baked into workload cost       | Yes — visible as the gap between the two costs 
```

| Identifies spare capacity    | Yes — explicit idle line item                 | Waste and idle merged into one unattributed pool    |

Neither is wrong — they answer different questions. Allocation-based answers *"who caused this bill?"* Usage-only answers *"who consumed what?"* The over-provisioning gap is the key distinction: allocation-based assigns it to the workload responsible, usage-only leaves it unattributed alongside genuinely idle capacity.

### Example:

  In this example we focus only on CPU costs to simplify the discussion.

  A cluster has one node provisioned with 8 CPUs at **$10/hr total**. Two workloads are running:

>   - namespace/app-a: requested 3 CPUs out of 8 available → pays for 3/8 = 37.5% of CPU

Reviewer: In this example, suppose, `app-a` allocates `3` CPUs for `10` mins, uses `1` CPU, and de-allocates all `3`. It uses the same 37.5% of CPU, but only for `10` mins. Does it still pay 37.5% of CPU?

2. **Wasted** — resources a workload reserved but did not use (over-provisioning) 
3. **Idle** — resources that no workload reserved at all — spare cluster capacity
4. **Shared** - resources that are common to all workloads - [discussed here](#shared-costs)

Note: There are cases of under-provisioning which result in efficiency (wasted) being greater than 100%. 

### Allocation-based cost — max(request, usage)

Each workload pays for what it reserved or what it actually uses, whichever is greater. If App-a reserved 3 CPUs and used 1 — it pays for 3. The wasted 2 CPUs are baked into App-a's bill. The idle CPUs (unreserved by anyone) appear as a separate `__idle__` line item. In another example, App-b requests 2 CPUs, and then bursts above this to use 4 CPUs, 4 CPUs will be attributed to App-b's costs. The cluster bill always reconciles — every dollar is accounted for across workloads plus idle.

### Usage-only cost

Each workload pays only for what it actually consumed. App-a pays for 1 CPU, same as app-b. The wasted 2 CPUs (app-a's over-provisioning) now join the idle pool as unattributed cost. **The numbers no longer add up to your overall bill.**

> ### The practical difference - usage vs reservation based costs

Reviewer 2: ```suggestion
### The practical difference - usage vs allocation based costs
```

> |                              | Reservation-based                             | Usage-only                                          |

Reviewer 2: ```suggestion
|                              | Allocation-based                             | Usage-only                                          |
```

|------------------------------|-----------------------------------------------|-----------------------------------------------------|
| Reconciles to cloud bill     | Yes                                           | No — waste + idle are unattributed                  |

> | Fair for team chargeback     | No — penalizes over-provisioning              | Yes                                                 |

Reviewer 2: I guess this line item is subjective if it is 'fair' - typically, the app team dictates the resource requirements of their app. Therefore, in this context, what is 'fair'? What we typically see is engineering groups grossly over provisioning their deployments to avoid outages in production. While that does indeed accomplish that goal, that creates a lot of waste and underutilized GPU. Wouldn't the fair outcome for a team to be charged for the resources they say they need, rather than what they use? it still costs the business the same amount of money whether they actually use what they request, so from a finance perspective, it is still a charge incurred by that cost center

| Incentivizes right-sizing    | Yes — you pay for what you reserve            | Less so                                             |
| Identifies over-provisioning | Not directly — baked into workload cost       | Yes — visible as the gap between the two cost figures |
| Identifies spare capacity    | Yes — explicit idle line item                 | Waste and idle merged into one unattributed pool    |

> Neither is wrong — they answer different questions. Reservation-based answers *"who caused this bill?"* Usage-only answers *"who consumed what?"* The over-provisioning gap is the key distinction: reservation-based assigns it to the workload responsible, usage-only leaves it unattributed alongside genuinely idle capacity.

Reviewer 2: ```suggestion
Neither is wrong — they answer different questions. Allocation-based answers *"who caused this bill?"* Usage-only answers *"who consumed what?"* The over-provisioning gap is the key distinction: allocation-based assigns it to the workload responsible, usage-only leaves it unattributed alongside genuinely idle capacity.
```

### Example:

  In this example we focus only on CPU costs to simplify the discussion.

  A cluster has one node provisioned with 8 CPUs at **$10/hr total**. Two workloads are running:

  - namespace/app-a: requested 3 CPUs out of 8 available → pays for 3/8 = 37.5% of CPU
  - namespace/app-b: requested 1 CPU → pays for 1/8 = 12.5% of CPU
  - 4 CPUs are idle (nobody requested them) → 50% of CPU

>   A cluster's CPU Cost represents the cost of usage + waste.

Reviewer 2: ```suggestion
  A cluster's CPU Cost represents the cost of usage + idle (waste).
```

  CPU Usage Cost is only the usage without taking into account the waste or idle.

>   For idle, there are two ways to attribute the cost of the idle CPUs.  Either across workloads or as a separate line item.

Reviewer 2: ```suggestion
  For idle, there are two ways to attribute the cost of the idle CPUs - either across workloads, or as a separate line item.
```

  **Separating out Idle:**

  

>   | Workload   | Requested | Used | Billed cores | CPU cost (of $10/hr) | CPU Usage Cost | Wasted |

Reviewer 2: ```suggestion
  | Workload   | Requested | Used | Billed cores | CPU cost (of $10/hr) | CPU Usage Cost | Idle |
```

  |------------|-----------|------|--------------|----------------------|----------------|--------|
  | app-a      | 3         | 1    | 3            | $3.75                | $1.25          | $2.50  |
  | app-b      | 1         | 1    | 1            | $1.25                | $1.25          | —      |
  | \_\_idle__ | —         | —    | 4            | **$5.00**            | —              | —      |

  **Weighted Distribution of Idle Costs:**

>   Each workload incorporates its relative percentage of the idle cost.  In our example 75% of the idle cost is attributed to app-a since it allocated 75% of the CPUs, and 25% of the idle cost is attributed to app-b since it allocated only one of the 4 CPUs.

Reviewer 2: ```suggestion
  Each workload incorporates its relative percentage of the idle cost.  In our example 75% of the idle cost is attributed to app-a since it accounted for 75% of total usage, and 25% of the idle cost is attributed to app-b since it comprised 25% of total usage.
```

Reviewer: In this example, it is not clear why the `4` idle CPUs in the cluster  should influence price for either `app-a` or `app-b`. If there were `4` idle CPUs during this hour, then no workload was starving for CPUs. The over-provisioning by `app-a` was immaterial, because no other workload needed these CPUs. If so, it's not clear to me what difference does this make to allocate idle costs for a very well behaving workload like `app-b` and penalizing `app-a` for over-allocation? Neither would improve the cluster ROI, would it?

>   | Workload | Requested | Used | Billed cores | CPU cost (of $10/hr) | CPU Usage Cost | Wasted |

Reviewer 2: ```suggestion
  | Workload | Requested | Used | Billed cores | CPU cost (of $10/hr) | CPU Usage Cost | Idle |
```

  |----------|-----------|------|--------------|----------------------|----------------|--------|
  | app-a    | 3         | 1    | 3            | **$7.50**            | $1.25          | $2.50  |
  | app-b    | 1         | 1    | 1            | **$2.50**            | $1.25          | —      |
  
    Usage only cost:         app-a=$1.25, app-b=$1.25, unattributed=$7.50  → total $10.00 ✓

>     Reservation based cost:  app-a=$3.75, app-b=$1.25, idle=$5.00  → total $10.00 ✓

Reviewer 2: these two lines might be able to be removed, they seem to just re-state info in the table. Maybe you could add a TOTALs row to the table if you wanted?
